### ISSUE TITLE
Fix smoke test path comparison logic, uncomment test which was broken [phab:D2323]

### ISSUE THREAD

**Issue by kythe** [google/kythe#2788]
_Created by **danielmoy-google** at 2018-05-18 23:49:53_:

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-05-19 23:12:02_:

```
rf := map[string][]byte{
   "foo.java": []byte("stuff"),
   "bar.java": []byte{"stuff"},
}
```

and in other places where this pattern is apropos

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-05-19 23:12:02_:

```
	cu := &kindex.Compilation{
		Proto: &apb.CompilationUnit{
			VName: &spb.VName{
				Corpus: "test",
				Path:   "fakepath",
			},
			SourceFile: files.requiredInput,
		},
	}
```

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-05-19 23:12:02_:

Deferred operations don't happen till the function returns, so the loop will stack them up. More importantly, we should check the error from closing a file that's open for writing, since some errors may not be observed till close.

Maybe pull this into a helper? (Warning: Untested example)

```
func writeUnitFile(ctx context.Context, path string, unit *kindex.Compilation) error {
   f, err := vfs.Create(ctx, path)
   if err != nil {
      return err
   } else if _, err := unit.WriteTo(f); err != nil {
      f.Close()
      return err
   }
   return f.Close()
}
```

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-05-19 23:12:02_:

The context should come in as a parameter.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-05-19 23:12:02_:

Please add a context parameter.

Be careful about APIs with lots of consecutive string-type arguments; it's really easy to mix up which is which at the call site, and the compiler can't help debug them. 

This case isn't so bad, but if it grows you might consider a parameter struct instead:


```
type Repo struct {
   URI        string
   OutputPath string
   ConfigPath string
}

func (e testExtractor) ExtractRepo(ctx context.Context, r Repo) error {
```

or words to that effect. This makes call sites self-documenting,

```
ext.ExtractRepo(ctx, smoke.Repo{
   URI:        "https://github.com/here/comes/the/airplane.git",
   OutputPath: "/tmp/repojunk/foo",
   ConfigPath: "/tmp/test/config.json",
})
```

**Comment by kythe** [comment]
_**fromberger** wrote at 2018-05-19 23:12:02_:

Can you please update the summary to be a bit more specific about what the refactoring is?
It looks like functional changes are included, and I have no real objection to that, but I wonder if we shouldn't separate those from true refactorings.

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-05-21 16:49:24_:

Consider renaming coverage to either: expectedCoverage or wantCoverage, to make it clear from the signature what the intention of the variable is.

Separately, consider making the expected value a struct {Extracted, Downloaded, wantCoverage} so that you can test for cases of expected failed fetching and/or extraction as well.

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-05-21 16:51:07_:

It might be worthwhile to check to see what the convention for CU paths are and test for correctness there as well.

**Comment by kythe** [comment]
_**danielmoy-google** wrote at 2018-05-21 17:14:58_:

>>! In (earlier comment), fromberger wrote:
> Can you please update the summary to be a bit more specific about what the refactoring is?
> It looks like functional changes are included, and I have no real objection to that, but I wonder if we shouldn't separate those from true refactorings.

Oh yea sorry, I was getting a bit chain-happy on the commits late last week and was basically just reusing the same env for all the commits, which means every commit beyond the first one had a nonsensical commit message.  Updated.

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-05-21 17:27:01_:

Noted, though for now this test harness has no dep on vname/cu Path at all, so it's explicitly not tested.  I imagine this could change in the future.

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-05-21 17:27:01_:

noted this refactor seems good to do, though I would prefer to do it in follow-up commit

**Comment by kythe** [comment]
_**danielmoy-google** wrote at 2018-05-21 17:28:08_:

>>! In (earlier comment), fromberger wrote:
> Can you please update the summary to be a bit more specific about what the refactoring is?
> It looks like functional changes are included, and I have no real objection to that, but I wonder if we shouldn't separate those from true refactorings.

Ah yea I got a bit lax with the chained commits from late last week, and forgot to updated the summaries.  The 'refactoring' was the first commit in a chain of like 4 commits, the other ~three (including this one) should have used different summaries, woops.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-05-21 17:46:27_:

That's reasonable. Please use `context.TODO()` instead of `context.Background()` in the places where it remains to be updated later—the result is the same, but it's easier to search for.

**Comment by kythe** [comment]
_**danielmoy-google** wrote at 2018-05-21 18:00:16_:

  - Merge branch 'master' into fuzz
  - Addressing comments

**Comment by kythe** [comment]
_**danielmoy-google** wrote at 2018-05-21 18:14:06_:

  - Fixed test, no longer using vfs

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-05-21 18:14:08_:

from offline discussion: probably shouldn't have been using vfs here in the first place, just using simple files.  Inside the unit test, all this is temp dir stuff anyways, so no point in taking the overhead that vfs brinsg.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-05-21 20:03:22_:

Please document what the parameters mean (particularly `expectedCoverage`)

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-05-21 20:03:22_:

`t.Helper()`

**Comment by kythe** [comment]
_**fromberger** accepted this revision at 2018-05-21 20:03:22_

**Comment by kythe** [comment]
_**danielmoy-google** closed via commit 98ec95c98296c552488ab978f17d3df7e8f6baa9 at 2018-05-21 20:08:49_
